onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,719 @@
+  for (NodeArg* input : require_type_change) {
+    if (IsType(*input, TensorProto::FLOAT)) {
+      input->UpdateTypeAndShape(type_proto, true, true, logger);
+    }
+  }
+}
+
+// PropagateForwards
+// Propagate FP32 Cast operations forwards (downstream)
+// Recurrsively search the graph for Cast FP16 safe/allowed operations to expand
+// the float16 computation region.
+// The required_cast vector is the collection of nodes that require float cast.
+// All nodeargs on a path down to any of the 
+// frontier nodes require type change from FLOAT to FLOAT16. 
+// require_type_change consists of such nodes.  All the frontier nodes require fp32 cast
+// The input node is expected to be non-nullptr
+static bool PropagateForwards(Graph& graph, Node* node,
+                              std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                              size_t level,
+                              const logging::Logger& logger)
+{
+  ORT_ENFORCE(node != nullptr);
+  bool modified = false;
+  if (IsCastTo(*node, TensorProto::FLOAT)) {
+    std::unordered_set<NodeArg*> require_cast;
+    std::unordered_set<NodeArg*> require_type_change;
+    NodeArg* cast_output = node->MutableOutputDefs()[0];
+    SearchDownstream(graph, cast_output, require_cast, require_type_change, removed_nodes, level);
+    if (require_cast.size() > 0 && require_cast.find(cast_output) == require_cast.end()) {
+      // Remove Cast operation
+      VLOGS(logger, 1) << "PropagateForwards: Removed Cast node  " << node->Name();
+      RemoveCastNodes(graph, {node}, removed_nodes);
+      InsertCastNodes(graph, require_cast, false, removed_nodes);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateForwwards: Inserted Cast nodes " << ConcatNames<std::unordered_set<NodeArg*>>(require_cast);
+      modified = true;
+    }
+  } else if (node->OpType() != "Cast") {
+    for (NodeArg* output: node->MutableOutputDefs()) {
+      for (Node* consumer : graph.GetMutableConsumerNodes(output->Name())) {
+        if (consumer && std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end()) {
+          modified |= PropagateForwards(graph, consumer, removed_nodes, level, logger);
+        }
+      }
+    }
+  }
+  return modified;
+}
+
+// PropagateBackwards
+// Propagate FP16 Cast operations backwards (upstream)
+// Recurrsively search the graph for Cast FP16 safe/allowed operations and expand
+// float16 computation regsion and
+// find the frontiers of the float16 computation region.
+// The required_cast vector is the collection of 
+// FP16-cast-frontiers of the cast node. All nodeargs on the path from any of the 
+// frontier nodes to the cast node require type change from  FLOAT to FLOAT16.
+// Each of the frontier nodes requires an fp16 cast.
+// The input node is expected be non-nullptr.
+static bool PropagateBackwards(Graph& graph, Node* node,
```

Comment:
we need a way to track visited nodes. 
I don't fully understand how this works... but I feel PropagateBackwards() has been called more times than needed repeatedly on the same node... ? 

PropagateBackwards() is first called for every node in the graph.Nodes() in the ApplyImpl. 
and then it's recursively called on every inputs of non-Cast nodes? 

If we want to recursive it backward, we should have a unique set of starting point. But every node in this graph is now a starting point?

Reply:
Please see the updated version.